specs/agents/tracing-spans-compress.md

```diff
@@ -0,0 +1,210 @@
+    }
+    compressed.exactMatch = compressed.exactMatch && exactMatch
+    endTimestamp = max(endTimestamp, other.endTimestamp)
+}
+```
+
+#### Effects on metric processing
+
+As laid out in the [span destination spec](tracing-spans-destination.md#contextdestinationserviceresource),
+APM Server tracks span destination metrics.
+To avoid compressed spans to skew latency metrics and cause throughput metrics to be under-counted,
+APM Server will take `composite.count` into account when tracking span destination metrics.
+
+### Compression algorithm
+
+#### Eligibility for compression
+
+A span is eligible for compression if all the following conditions are met
+- It's an [exit span](https://github.com/elastic/apm/blob/master/specs/agents/tracing-spans-destination.md#contextdestinationserviceresource)
+- The trace context of this span has not been propagated to a downstream service
+
+The latter condition is important so that we don't remove (compress) a span that may be the parent of a downstream service.
+This would orphan the sub-graph started by the downstream service and cause it to not appear in the waterfall view.
+
+```java
+boolean isCompressionEligible() {
+    return exit && !context.hasPropagated
+}
+```
+
+#### Span buffering
+
+Non-compression-eligible spans may be reported immediately after they have ended.
+When a compression-eligible span ends, it does not immediately get reported.
+Instead, the span is buffered within its parent.
+A span/transaction can buffer at most one child span.
+
+Span buffering allows to "look back" one span when determining whether a given span should be compressed.
+
+A buffered span gets reported when
+1. its parent ends
+2. a non-compressible sibling ends
+
+```java
+void onSpanEnd() {
+    if (isCompressionEligible()) {
+        if (parent.hasBufferedSpan()) {
+            parent.tryCompress(this)
+        } else {
+            parent.buffered = this
+        }
+    } else { 
+        report(buffered)
+        report(this)
+    }
+}
+```
+
+#### Compression
+
+On span end, we compare each [compression-eligible](tracing-spans-compress.md#eligibility-for-compression) span to it's previous sibling.
+
+If the spans are of the same kind but have different span names and the compressions-eligible span's `duration` <= `same_kind_compression_max_duration`,
+we compress them using the [Consecutive-Same-Kind compression strategy](tracing-spans-compress.md#consecutive-same-kind-compression-strategy).
+
+If the spans are of the same kind, and have the same name,
+we apply the [Consecutive-Exact-Match compression strategy](tracing-spans-compress.md#consecutive-exact-match-compression-strategy).
+
+```java
+void tryCompress(Span child) {
+    if (buffered.isSameKind(child)) {
+        if (buffered.name == child.name) {
+            buffered.compress(child, exactMatch: true)
+            return
+        } else if ( (buffered.duration <= same_kind_compression_max_duration || buffered.composite.count > 1)
```

Comment:
Does `buffered.duration` refer to `duration` of the composite span? Could you please clarify why `buffered.duration <= same_kind_compression_max_duration` condition is necessary?

Reply:
`buffered` may contain either a regular or a composite span. If `buffered.composite.count > 1`, we're talking about a composite span.
For composite spans, we don't need to look at the condition `buffered.duration <= same_kind_compression_max_duration`, as the max duration only applies to regular (non-composite) spans.

Reply:
So IIUC with this condition is there to handle the case when the first span is buffered without checking its duration because if the second span is the exact match then we can compress them without considering duration but if the second span is not the exact match but it's the same kind then we need to go back to the first span and check its duration to allow same-kind compression. Did I understand the intent correctly?

OTOH for example if the first two spans are the exact match (and one or both of them might have duration larger than `same_kind_compression_max_duration`) and then the third span is not exact match but it's the same kind then strategy is switched to _same-kind_ but the algorithm doesn't go back to re-check that all the spans that were compressed when the strategy was _exact-match_ have duration not larger than `same_kind_compression_max_duration`.

Maybe it would simpler to not support strategy switch at all? The simpler approach might be that the strategy is fixed by the first two spans and it cannot be changed. When the `child` span cannot be appended to `buffered` while maintaining the selected strategy invariants then `buffered` span is sent out and the new one is started with `child` as its initial content.

Reply:
> Did I understand the intent correctly?

Yes, that's correct

You're absolutely right on the issue that can happen when the strategy changes, nice catch! 

> Maybe it would be simpler to not support strategy switch at all?

That sounds like a good option. We'll then have to store the applied strategy, maybe in `buffered.composite.strategy=[same-kind-match|exact-match]`

Reply:
@SergeyKleyman could you change the pseudo-code to reflect your suggestion?

Reply:
> @SergeyKleyman could you change the pseudo-code to reflect your suggestion?

Done - please take a look